src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -1479,34 +1457,29 @@ public override void WriteJson(JsonWriter writer, Stream? value, JsonSerializer
         [DebuggerDisplay("{" + nameof(DebuggerDisplay) + "}")]
         private class RpcMarshalableImplicitConverter : JsonConverter
         {
-            private readonly Type implicitlyConvertedType;
             private readonly JsonMessageFormatter jsonMessageFormatter;
-            private readonly JsonRpcProxyOptions proxyOptions;
-            private readonly JsonRpcTargetOptions targetOptions;
 
-            public RpcMarshalableImplicitConverter(Type implicitlyConvertedType, JsonMessageFormatter jsonMessageFormatter, JsonRpcProxyOptions proxyOptions, JsonRpcTargetOptions targetOptions)
+            public RpcMarshalableImplicitConverter(JsonMessageFormatter jsonMessageFormatter)
             {
-                this.implicitlyConvertedType = implicitlyConvertedType;
                 this.jsonMessageFormatter = jsonMessageFormatter;
-                this.proxyOptions = proxyOptions;
-                this.targetOptions = targetOptions;
-            }
-
-            private RpcMarshalableImplicitConverter(RpcMarshalableImplicitConverter copyFrom, Type implicitlyConvertedType)
-                : this(implicitlyConvertedType, copyFrom.jsonMessageFormatter, copyFrom.proxyOptions, copyFrom.targetOptions)
-            {
             }
 
-            private string DebuggerDisplay => $"Implicit converter for: {this.implicitlyConvertedType.Name}";
+            private string DebuggerDisplay => $"Converter for marhalable interfaces";
 
             public override bool CanConvert(Type objectType) =>
-                objectType == this.implicitlyConvertedType ||
-                (this.implicitlyConvertedType.IsGenericTypeDefinition && objectType.IsGenericType && objectType.GetGenericTypeDefinition() == this.implicitlyConvertedType);
+                MessageFormatterRpcMarshaledContextTracker.GetMarshalOptionsForType(objectType) is not null;
 
             public override object? ReadJson(JsonReader reader, Type objectType, object? existingValue, JsonSerializer serializer)
             {
+                (JsonRpcProxyOptions ProxyOptions, JsonRpcTargetOptions TargetOptions)? marshalOptions
```

Comment:
The current implementation is simpler than the original one which had multiple instances of `RpcMarshalableImplicitConverter`, one for each marshalable type. On the other hand, we are not caching the options anymore and we have to evaluate the `objectType` twice (one during `CanConvert` and then again during `ReadJson`). Should I go back to caching the options in different instances of `RpcMarshalableImplicitConverter` (one per type)?

Reply:
I think what you have here is fine, so long as if it's too slow to look up the options for a given type, we can optimize the `GetMarshalOptionsForType` method itself to cache the result.